src/snowflake/snowpark/modin/plugin/extensions/datetime_index.py

```diff
@@ -0,0 +1,124 @@
+            Datetime-like data to construct index with.
+        freq : str or pandas offset object, optional
+            One of pandas date offset strings or corresponding objects. The string
+            'infer' can be passed in order to set the frequency of the index as the
+            inferred frequency upon creation.
+        tz : pytz.timezone or dateutil.tz.tzfile or datetime.tzinfo or str
+            Set the Timezone of the data.
+        normalize : bool, default False
+            Normalize start/end dates to midnight before generating date range.
+        closed : {'left', 'right'}, optional
+            Set whether to include `start` and `end` that are on the
+            boundary. The default includes boundary points on either end.
+        ambiguous : 'infer', bool-ndarray, 'NaT', default 'raise'
+            When clocks moved backward due to DST, ambiguous times may arise.
+            For example in Central European Time (UTC+01), when going from 03:00
+            DST to 02:00 non-DST, 02:30:00 local time occurs both at 00:30:00 UTC
+            and at 01:30:00 UTC. In such a situation, the `ambiguous` parameter
+            dictates how ambiguous times should be handled.
+
+            - 'infer' will attempt to infer fall dst-transition hours based on
+              order
+            - bool-ndarray where True signifies a DST time, False signifies a
+              non-DST time (note that this flag is only applicable for ambiguous
+              times)
+            - 'NaT' will return NaT where there are ambiguous times
+            - 'raise' will raise an AmbiguousTimeError if there are ambiguous times.
+        dayfirst : bool, default False
+            If True, parse dates in `data` with the day first order.
+        yearfirst : bool, default False
+            If True parse dates in `data` with the year first order.
+        dtype : numpy.dtype or DatetimeTZDtype or str, default None
+            Note that the only NumPy dtype allowed is `datetime64[ns]`.
+        copy : bool, default False
+            Make a copy of input ndarray.
+        name : label, default None
+            Name to be stored in the index.
+
+        Examples
+        --------
+        >>> idx = pd.DatetimeIndex(["1/1/2020 10:00:00+00:00", "2/1/2020 11:00:00+00:00"], tz="America/Los_Angeles")
+        >>> idx
+        DatetimeIndex(['2020-01-01 02:00:00-08:00', '2020-02-01 03:00:00-08:00'], dtype='datetime64[ns, America/Los_Angeles]', freq=None)
+        """
+        kwargs = {
+            "freq": freq,
+            "tz": tz,
+            "normalize": normalize,
+            "closed": closed,
+            "ambiguous": ambiguous,
+            "dayfirst": dayfirst,
+            "yearfirst": yearfirst,
+            "dtype": dtype,
+            "copy": copy,
+            "name": name,
+        }
+        self.is_lazy = convert_to_lazy
+        if self.is_lazy:
+            self.set_query_compiler(data=data, **kwargs)
```

Comment:
if the original data is already a query compiler, shall we aways call a to_timestamp to convert it to timestamp type to make the index valid?

Reply:
I think we should not do unnecessary casting. For correctness we can add check that query_compiler passed here must be of type datetime.

Reply:
> i think it would be better to have consistent behavior, we either error out for both cases or allow it for both cases.

I didn't understand what 2 cases you are referring to. Can you please provide examples?

Reply:
Right now, if we do pd.DatetimeIndex([1, 2, 3]), this works fine, because we call native_pd.DatetimeIndex([1, 2, 3]) with will convert the integer type to datetime to create a valid DatetimeIndex, and then upload the data to snowflake.
However, if we have an integer column "A" in snowflake, and we create a DateTimeIndex out of that column, it raises an error.
@sfc-gh-azhan i think we have discussed that offline, and we think if user is calling DateTimeIndex constructer on an integer or string column, it is like saying they want to convert the column to timestamp column to construct a valid DatetimeIndex, and we can follow up to make this work in the next pr.
@sfc-gh-nkumar for now can we raise a NotImplementedError instead of ValueError, and add a TODO to follow this up.

Reply:
More specifically, those cases are not well supported in this PR IMO:
`pd.DatetimeIndex(pd.Series([1,2,3]))` or `pd.DatetimeIndex(pd.Index([1,2,3]))`.
We could put those cases in next PR and reuse `pd.to_datetime()` for them.

Reply:
First case is interesting because with Varnikas recent change series passed in will be treated as parent not actual data, we might have to rethink our design of passing parent to Index. 
I will add a failing test for second test with a follow up JIRA.

Reply:
follow up JIRA https://snowflakecomputing.atlassian.net/browse/SNOW-1616989